dart/lib/src/client_reports/client_report_recorder.dart

```diff
@@ -0,0 +1,53 @@
+import 'package:meta/meta.dart';
+
+import '../utils.dart';
+import 'client_report.dart';
+import 'discarded_event.dart';
+import 'outcome.dart';
+import '../transport/rate_limit_category.dart';
+
+@internal
+class ClientReportRecorder {
+  ClientReportRecorder([this._dateTimeProvider = getUtcDateTime]);
+
+  final DateTime Function() _dateTimeProvider;
+  final Map<_QuantityKey, int> _quantities = {};
+
+  void recordLostEvent(final Outcome reason, final RateLimitCategory category) {
+    final key = _QuantityKey(reason, category);
+    var current = _quantities[key] ?? 0;
+    _quantities[key] = current + 1;
+  }
+
+  ClientReport? flush() {
+    if (_quantities.isEmpty) {
+      return null;
+    }
+
+    final events = _quantities.keys.map((key) {
+      final quantity = _quantities[key] ?? 0;
```

Comment:
Should we just discard this item if zeroed?

Reply:
This is more of a code fallback, as it is not possible that this is nil, as we are iterating the keys and never set a key with empty value.